Add the correct nullable annotations to generated iterator code

chsienki opened the pull request:
This is mainly about plumbing through `TypeSymbolWithAnnotations` to the correct point so we can generate the attribute based on it's value.

- Expose the `IteratorElementType` as a `TypeSymbolWithAnnotations`
- Return the annotated type, or a default as needed from the required locations
- Allow state machine field to take a TSWA and pass it in via the iterator rewriter
- Fix test

Fixes #30010

chsienki commented:
@dotnet/roslyn-compiler for review please

cston commented on src/Compilers/CSharp/Portable/Lowering/MethodToClassRewriter.cs:
```diff
@@ -672,12 +672,12 @@ internal override void AddSynthesizedAttributes(PEModuleBuilder moduleBuilder, r
                 {
                     // BaseMethodWrapperSymbol should not be rewritten by the IteratorRewriter
-                    return null;
+                    return TypeSymbolWithAnnotations.Create(null);
```
>TypeSymbolWithAnnotations.Create(null) [](start = 27, length = 38)

`default`

jaredpar replied:
Rather that `default` should we add a `static` property with a decent name: Missing, Null, etc ...

cston commented on src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:
```diff
@@ -900,17 +900,17 @@ internal bool IsIterator
+        internal virtual TypeSymbolWithAnnotations IteratorElementType
         {
-            get { return null; }
+            get { return TypeSymbolWithAnnotations.Create(null); }
```
>TypeSymbolWithAnnotations.Create(null) [](start = 25, length = 38)

`default`

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs:
```diff
@@ -289,16 +289,16 @@ public override bool IsExtensionMethod
-                Debug.Assert((object)_iteratorElementType == null || TypeSymbol.Equals(_iteratorElementType, value, TypeCompareKind.ConsiderEverything2));
-                Interlocked.CompareExchange(ref _iteratorElementType, value, null);
+                Debug.Assert(_iteratorElementType.IsDefault || TypeSymbol.Equals(_iteratorElementType.TypeSymbol, value.TypeSymbol, TypeCompareKind.ConsiderEverything2));
+                _iteratorElementType = value;
```
>_iteratorElementType = value; [](start = 16, length = 29)

The assignment is not thread-safe.

`_iteratorElementType` should probably be a `TypeSymbolWithAnnotations.Builder`. 

Then this property would be:
```
get => _iteratorElementType.ToType();
set => _iteratorElementType.InterlockedInitialize(value);
```

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs:
```diff
@@ -706,16 +706,16 @@ internal sealed override bool TryGetThisParameter(out ParameterSymbol thisParame
-                Debug.Assert((object)_iteratorElementType == null || TypeSymbol.Equals(_iteratorElementType, value, TypeCompareKind.ConsiderEverything2));
-                Interlocked.CompareExchange(ref _iteratorElementType, value, null);
+                Debug.Assert(_iteratorElementType.IsDefault || TypeSymbol.Equals(_iteratorElementType.TypeSymbol, value.TypeSymbol, TypeCompareKind.ConsiderEverything2));
+                _iteratorElementType = value;
```
>_iteratorElementType = value [](start = 16, length = 28)

The assignment is not thread-safe.

chsienki commented:
@dotnet/roslyn-compiler for second review please

chsienki commented:
Rebase onto master and fixed up the errors.

I also renamed `IteratorElementType` to `IteratorElementTypeWithAnnotations` to be consistent with @gafter's changes
